src/Storages/ObjectStorage/StorageObjectStorageStableTaskDistributor.cpp

```diff
@@ -0,0 +1,175 @@
+#include "StorageObjectStorageStableTaskDistributor.h"
+#include <Common/SipHash.h>
+#include <consistent_hashing.h>
+#include <optional>
+
+namespace DB
+{
+
+StorageObjectStorageStableTaskDistributor::StorageObjectStorageStableTaskDistributor(
+    std::shared_ptr<IObjectIterator> iterator_,
+    std::optional<std::vector<std::string>> ids_of_nodes_)
+    : iterator(std::move(iterator_))
+    , connection_to_files(ids_of_nodes_.has_value() ? ids_of_nodes_.value().size() : 1)
+    , ids_of_nodes(ids_of_nodes_)
+    , iterator_exhausted(false)
+{
+}
+
+std::optional<String> StorageObjectStorageStableTaskDistributor::getNextTask(size_t number_of_current_replica)
+{
+    LOG_TRACE(
+        log,
+        "Received a new connection from replica {} looking for a file",
+        number_of_current_replica
+    );
+
+    // 1. Check pre-queued files first
+    if (auto file = getPreQueuedFile(number_of_current_replica))
+        return file;
+
+    // 2. Try to find a matching file from the iterator
+    if (auto file = getMatchingFileFromIterator(number_of_current_replica))
+        return file;
+
+    // 3. Process unprocessed files if iterator is exhausted
+    return getAnyUnprocessedFile(number_of_current_replica);
+}
+
+size_t StorageObjectStorageStableTaskDistributor::getReplicaForFile(const String & file_path)
+{
+    if (!ids_of_nodes.has_value())
```

Comment:
In which case would it not have a value? And why is `empty != nullopt`?

Reply:
It must always have value, in `ReadFromCluster::createExtension` it is called with `ids_of_nodes` always.

About nullopt - it is the same:
```
if (var)
if (var.has_value())
if (var != std::nullopt)
```
First and second are synonyms - https://en.cppreference.com/w/cpp/utility/optional/operator_bool
Third - because comparation with nullopt calls bool operator:
```
/usr/include/c++/11$ grep -B 5 -A 5 'operator==' optional
...
  // Comparisons with nullopt.
  template<typename _Tp>
    constexpr bool
    operator==(const optional<_Tp>& __lhs, nullopt_t) noexcept
    { return !__lhs; }
...
  template<typename _Tp>
    constexpr bool
    operator==(nullopt_t, const optional<_Tp>& __rhs) noexcept
    { return !__rhs; }
```

Reply:
Sorry, I didn't explain my question correctly. I was asking why not just have a vector and check if it is empty instead of an optional of vector

Reply:
What I mean is that instead of doing:

```
std::optional<std::vector<std::string>> ids_of_nodes;
...
    if (!ids_of_nodes.has_value())
```


you simply do:

```
std::vector<std::string> ids_of_nodes;
...
    if (ids_of_nodes.empty())
```

I mean, it is ok to keep it as optional, just thought it could be simpler